src/Framework/FieldsAPI/Concerns/HasPersistence.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace Give\Framework\FieldsAPI\Concerns;
+
+use Give\Framework\FieldsAPI\ValueObjects\PersistenceScope;
+
+/**
+ * This provides the ability to set a scope and meta key for a field. The scope is used to determine if and where the
+ * field should be stored. The meta key is used to store the field in the database.
+ *
+ * @unreleased
+ */
+trait HasPersistence
+{
+    /**
+     * @unreleased
+     *
+     * @var PersistenceScope|null
+     */
+    protected $scope;
+
+    /**
+     * @unreleased
+     *
+     * @var string|null
+     */
+    protected $metaKey = null;
+
+    /**
+     * @unreleased
+     *
+     * @param string|PersistenceScope $scope
+     */
+    public function scope($scope): self
+    {
+        $this->scope = $scope instanceof PersistenceScope ? $scope : new PersistenceScope($scope);
+
+        return $this;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getScope(): PersistenceScope
+    {
+        return $this->scope;
+    }
+
+    /**
+     * @unreleased
+     */
+    public function getScopeValue(): string
```

Comment:
Let me know if i'm understanding how we would implement this...

```
$field->scope('my_custom_scope');
```

Donation Form Context
```
foreach($fields as $field){
  $scope = $field->getScopeValue();
  
  do_action('givewp_donation_form_persistence_scope_$scope}', $value, $donation);
}
```

```
add_action('givewp_donation_form_persistence_scope_my_custom_scope', function($value, $donation){
  // do something
});
```

Some Other Context like options?
```
foreach($fields as $field){
  $scope = $field->getScopeValue();
  
  do_action('givewp_options_persistence_scope_$scope}', $value);
}
```

```
add_action('givewp_options_persistence_scope_my_custom_scope', function($value){
  // do something
});
```

Reply:
@jonwaldstein Yeah, for our donation form persistence, we'll have something like:
```php
public function storeCustomField(Field $field, $value) {
	switch($field->getScopeValue()) {
		case 'donation':
			// store donation meta
			break;

		case 'donor':
			// store donor meta
			break;

		default:
			do_action("givewp_donation_form_persist_field_scope_{$field->getScopeValue()}", $field, $value);
	}
}
```

Reply:
@JasonTheAdams okay yeah that make sense. It's certainly flexible - my main hesitation is that this has the potential of being _too_ flexible - meaning someone can find the name of your intended scope and override it either knowingly or unknowingly.  The scope "custom" is the first thing that comes to mind.  Imagine folks with different fields trying to make custom scopes and each one is using the word "custom" - it would be quite confusing when they don't work properly. Being this is directly related to persisting the field, we want it to be flexible - but not have the potential for the processing to stop working because of a conflict like this.  You could of course argue that it's more of a responsibility on the implementation like the hook name, etc..

One other idea I was thinking about for this api is more like how the validation rules work.  Either you could use the build in strings `$field->scope('donation')` or instead, provide a custom invokable class `$field->scope(new FieldScope())`. Those scopes could implement a simple contract that would receive the `$field`, `$value`, and `$context`/`$data` (extra request data like the Donation model).  I know it sounds a little complex but I like the idea of scopes being more like a singleton.

Reply:
I've added the file and callback scopes per our discussion. 85dcb734d